autogpt_platform/autogpt_libs/autogpt_libs/auth/jwt_utils_test.py

```diff
@@ -0,0 +1,391 @@
+        """Test JWT token with additional custom claims."""
+        mocker.patch.dict(
+            os.environ,
+            {"JWT_VERIFY_KEY": self.valid_secret},
+            clear=True,
+        )
+        from autogpt_libs.auth import jwt_utils
+
+        mocker.patch.object(jwt_utils, "settings", Settings())
+        extra_claims_payload = {
+            "sub": "user-id",
+            "role": "user",
+            "aud": "authenticated",
+            "custom_claim": "custom_value",
+            "permissions": ["read", "write"],
+            "metadata": {"key": "value"},
+        }
+        token = jwt.encode(extra_claims_payload, self.valid_secret, algorithm="HS256")
+
+        result = parse_jwt_token(token)
+        assert result["sub"] == "user-id"
+        assert result["custom_claim"] == "custom_value"
+        assert result["permissions"] == ["read", "write"]
+
+    def test_jwt_with_numeric_sub(self):
+        """Test JWT token with numeric user ID."""
+        payload = {
+            "sub": 12345,  # Numeric ID
+            "role": "user",
+            "aud": "authenticated",
+        }
+        # Should convert to string internally
+        user = verify_user(payload, admin_only=False)
+        assert user.user_id == 12345
+
+    def test_jwt_with_very_long_sub(self):
+        """Test JWT token with very long user ID."""
+        long_id = "a" * 1000
+        payload = {
+            "sub": long_id,
+            "role": "user",
+            "aud": "authenticated",
+        }
+        user = verify_user(payload, admin_only=False)
+        assert user.user_id == long_id
+
+    def test_jwt_with_special_characters_in_claims(self):
+        """Test JWT token with special characters in claims."""
+        payload = {
+            "sub": "user@example.com/special-chars!@#$%",
+            "role": "admin",
+            "aud": "authenticated",
+            "email": "test+special@example.com",
+        }
+        user = verify_user(payload, admin_only=True)
+        assert "special-chars!@#$%" in user.user_id
+
+    def test_jwt_with_future_iat(self, mocker: MockerFixture):
+        """Test JWT token with issued-at time in future."""
+        mocker.patch.dict(
+            os.environ,
+            {"JWT_VERIFY_KEY": self.valid_secret},
+            clear=True,
+        )
+        from autogpt_libs.auth import jwt_utils
+
+        mocker.patch.object(jwt_utils, "settings", Settings())
+        future_payload = {
+            "sub": "user-id",
+            "role": "user",
+            "aud": "authenticated",
+            "iat": datetime.now(timezone.utc) + timedelta(hours=1),
+        }
+        token = jwt.encode(future_payload, self.valid_secret, algorithm="HS256")
+
+        # PyJWT validates iat claim and should reject future tokens
+        with pytest.raises(ValueError, match="not yet valid"):
+            parse_jwt_token(token)
+
+    def test_jwt_with_different_algorithms(self, mocker: MockerFixture):
```

Comment:
Add a constant time test maybe?

Reply:
I have checked the source code and we're all good. Writing constant-time tests involving `jwt.decode` doesn't really work because it also decodes the JSON which is not constant-time, even if the signature verification is.